### ISSUE TITLE
[core] center still offset incorrectly with content insets

### ISSUE THREAD

**Issue by chloekraw** [mapbox/mapbox-gl-native#15344]
Customer reported a bug with the center being offset incorrectly when setting the camera with the `release-oolong` branch. Likely related to https://github.com/mapbox/mapbox-gl-native/issues/15106, maybe https://github.com/mapbox/mapbox-gl-native/issues/14991. 

We expected https://github.com/mapbox/mapbox-gl-native/pull/15130 to resolve this regression but this fix has already been backported to `oolong` in https://github.com/mapbox/mapbox-gl-native/pull/15176.

Customer suspects it's from this part of their code:

```
  mbgl::CameraOptions cameraOptions;
  cameraOptions.center = MGLLatLngFromLocationCoordinate2D(centerCoordinate);
  cameraOptions.padding = MGLEdgeInsetsFromNSEdgeInsets(insets);
  cameraOptions.zoom = zoomLevel;
  if (direction >= 0) {
    cameraOptions.bearing = direction;
  }
```

Problems reported with the map:
- try to set the center of the map to the current location and notice the current location is slightly off-center
- tell the camera to position on the map bounds of a specific city. map bounds have the same offset issue
- center the map to current location. there's the same slightly off-center offset
- try again to center the map on current location. when it properly positions the map to center, the offset bug goes away.

**Comment by astojilj** [comment]
Based on the problem description, this code behaves as expected: padding moves all content, and center (and perspective center if pitch is used) to content area defined by padding.
Padding is also persistent property of camera - once it is set, it is active until other, or zero, padding is set.

#15130 was about offset in selecting annotations, it didn't affect how content is rendered (rendering offset).

Padding on iOS is used always because of top bar - so that the center is in visible area. Then, if there are other UI elements overlaid on sides, e.g. turn by turn navigation, padding is used to center the content to non occluded area of screen.

What might be causing issue is that, insets are defined in[ logical screen coordinates](https://docs.mapbox.com/android/api/map-sdk/8.1.0/com/mapbox/mapboxsdk/maps/MapboxMap.html#setPadding-int-int-int-int-), as opposite to using physical - the customer in this case might have just the vale that is too high. Without knowing the value set in code above, it is hard to tell.

**Comment by chloekraw** [comment]
Thanks @astojilj. I’ll email them and make sure they see it. 

Let’s say that they had this same code and the implementation back when they were on `release-mojito`. If they notice a difference in behavior now, why might that be?

**Comment by astojilj** [comment]
@chloekraw 

https://github.com/mapbox/mapbox-gl-native/pull/14664 introduced persistent nature of padding - earlier, set padding was reset to 0 after the transition and new center was calculated (padding was baked into center coordinate). As padding defines content area, asymmetric viewport center and vanishing point, it had to be set as persistent.

After #14664, center coordinate and padding are separate, and if there is a need to reset padding, there should be an explicit call from client code to do it. Note that padding, as other camera properties are, is animated when using easeTo/flyTo.

**Comment by chloekraw** [comment]
Thanks @astojilj. Closing this ticket, will re-open if it's needed.

**Comment by fabian-guerra** [comment]
I'm re-opening this issue because as of 5.0.0 setting `MGLMapView.contentInset` effectively moved `MGLMapView.centerCoordinate` accordingly. Stated in our documentation: 

https://github.com/mapbox/mapbox-gl-native/blob/39af6b48c52b1618d5b9fb51bcafad111ec01c7f/platform/ios/src/MGLMapView.h#L1306-L1310

Starting `5.1.0` this is no longer true making it a regression.

**Comment by astojilj** [comment]
>   When the value of this property is equal to `UIEdgeInsetsZero`, viewport 
>   properties such as `centerCoordinate` assume a viewport that matches the map 
>   view’s frame. Otherwise, those properties are inset, excluding part of the 
>   frame from the viewport. For instance, if the only the top edge is inset, the 
>   map center is effectively shifted downward.

@fabian-guerra @chloekraw @1ec5 Let's move this to another issue and not reopen this one, please.

@fabian-guerra that content above doesnt mean that value in centerCoordinate gets altered when you change contentInset. It is just rendered on different place.
